akka-bench-jmh/src/main/scala/akka/stream/FramingBenchmark.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream
+
+import java.util.concurrent.{ Semaphore, TimeUnit }
+
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.stream.scaladsl.{ Framing, Sink, Source }
+import akka.util.ByteString
+import com.typesafe.config.{ Config, ConfigFactory }
+import org.openjdk.jmh.annotations._
+
+import scala.concurrent.Await
+import scala.concurrent.duration._
+import scala.util.Random
+
+@State(Scope.Benchmark)
+@OutputTimeUnit(TimeUnit.SECONDS)
+@BenchmarkMode(Array(Mode.Throughput))
+class FramingBenchmark {
+
+  val config: Config = ConfigFactory.parseString(
+    """
+      akka {
+        log-config-on-start = off
+        log-dead-letters-during-shutdown = off
+        stdout-loglevel = "OFF"
+        loglevel = "OFF"
+        actor.default-dispatcher {
+          #executor = "thread-pool-executor"
+          throughput = 1024
+        }
+        actor.default-mailbox {
+          mailbox-type = "akka.dispatch.SingleConsumerOnlyUnboundedMailbox"
+        }
+        test {
+          timefactor =  1.0
+          filter-leeway = 3s
+          single-expect-default = 3s
+          default-timeout = 5s
+          calling-thread-dispatcher {
+            type = akka.testkit.CallingThreadDispatcherConfigurator
+          }
+        }
+      }""".stripMargin
+  ).withFallback(ConfigFactory.load())
+
+  implicit val system: ActorSystem = ActorSystem("test", config)
+
+  var materializer: ActorMaterializer = _
+
+  // Safe to be benchmark scoped because the flows we construct in this bench are stateless
+  var flow: Source[ByteString, NotUsed] = _
+
+  @Param(Array("32", "64", "128", "256", "512", "1024"))
+  var messageSize = 0
+
+  @Param(Array("1", "8", "16", "32", "64", "128"))
+  var framePerSeq = 0
+
+  @Setup
+  def setup(): Unit = {
+    materializer = ActorMaterializer()
+
+    val frame = List.range(0, messageSize, 1).map(_ ⇒ Random.nextPrintableChar()).mkString + "\n"
+    flow = Source.repeat(ByteString(List.range(0, framePerSeq, 1).map(_ ⇒ frame).mkString)).take(100000)
+      .via(Framing.delimiter(ByteString("\n"), Int.MaxValue))
+  }
+
+  @TearDown
+  def shutdown(): Unit = {
+    Await.result(system.terminate(), 5.seconds)
+  }
+
+  @Benchmark
+  @OperationsPerInvocation(100000)
+  def framing(): Unit = {
+    val lock = new Semaphore(1)
+    lock.acquire()
+    flow.runWith(Sink.onComplete(_ ⇒ lock.release()))(materializer)
+    lock.acquire()
+  }
```

Comment:
When I run this with `jmh:run -prof gc -i 5 -wi 3 -f1 -t1 akka.stream.FramingBenchmark` with a few of the parameters.

Before (master):
```
[info] Benchmark                                              (framePerSeq)  (messageSize)   Mode  Cnt       Score        Error   Units
[info] FramingBenchmark.framing                                          32            256  thrpt    5   53457.467 ±   2421.302   ops/s
[info] FramingBenchmark.framing:·gc.alloc.rate                           32            256  thrpt    5    5574.247 ±  12001.918  MB/sec
[info] FramingBenchmark.framing:·gc.alloc.rate.norm                      32            256  thrpt    5  114510.233 ± 246491.634    B/op
[info] FramingBenchmark.framing                                          64            256  thrpt    5   18186.856 ±   1083.036   ops/s
[info] FramingBenchmark.framing:·gc.alloc.rate                           64            256  thrpt    5    7323.029 ±  15768.183  MB/sec
[info] FramingBenchmark.framing:·gc.alloc.rate.norm                      64            256  thrpt    5  439606.035 ± 946284.571    B/op
```

After (this PR):
```
[info] Benchmark                                              (framePerSeq)  (messageSize)   Mode  Cnt       Score       Error   Units
[info] FramingBenchmark.framing                                          32            256  thrpt    5  107170.492 ±  2757.952   ops/s
[info] FramingBenchmark.framing:·gc.alloc.rate                           32            256  thrpt    5    1344.358 ±  2893.928  MB/sec
[info] FramingBenchmark.framing:·gc.alloc.rate.norm                      32            256  thrpt    5   13767.800 ± 29635.991    B/op
[info] FramingBenchmark.framing                                          64            256  thrpt    5   41255.045 ±  2386.223   ops/s
[info] FramingBenchmark.framing:·gc.alloc.rate                           64            256  thrpt    5    1044.139 ±  2247.898  MB/sec
[info] FramingBenchmark.framing:·gc.alloc.rate.norm                      64            256  thrpt    5   27490.697 ± 59175.469    B/op
```

That looks like a rather significant improvement!